File: Week2/convertSortedArrayToBinarySearchTree.js
var sortedArrayToBST = function(nums) {
    return dfs(0, nums.length - 1, nums);
};

function dfs(left, right, nums) {
    if (left > right) {
     return null;   
    }
    if (left === right) {
        return new TreeNode(nums[left])
    }
    let mid = Math.floor((left + right) / 2);
    let root = new TreeNode(nums[mid]);
    root.left = dfs(left, mid - 1, nums);
    root.right = dfs(mid + 1, right, nums);
    return root;
}

File: Week6/ConvertSortedArrayToBinarySearchTree.java
public class ConvertSortedArrayToBinarySearchTree {

    public TreeNode sortedArrayToBST(int[] nums) {
        if (nums.length == 0) {
            return null;
        }
        return construct(nums, 0, nums.length - 1);
    }

    private TreeNode construct(int[] nums, int lo, int hi) {
        int mid = lo + (hi - lo) / 2;
        TreeNode node = new TreeNode(nums[mid]);
        if (mid - 1 >= lo) {
            node.left = construct(nums, lo, mid - 1);
        }
        if (mid + 1 <= hi) {
            node.right = construct(nums, mid + 1, hi);
        }
        return node;
    }
}
